packages/shared-state/files/usr/lib/lua/shared-state.lua

```diff
@@ -156,6 +155,20 @@ local function SharedState(dataType, pLogger)
 		end
 	end
 
+	function sharedState.httpRequest(url, body)
+		local cmd = "uclient-fetch -q -O- --timeout=3 --post-data='"..body.."' '"..url.."'"
```

Comment:
Just remembered that Lua io.popen is unidirectional so either we read from it or we write to it but cannot do both, so we still need to use a temporary file instead of stdin here,

Reply:
> What happens if body is big or if it contains chars like `'` ?

@spiccinini's utils.shell_quote can be used for this maybe?
https://github.com/libremesh/lime-packages/blob/606ae4c6b02118571a3d312b6426d1c57e4e796c/packages/lime-system/files/usr/lib/lua/lime/utils.lua#L275-L279

Reply:
As per some tests, 65KB seems to be the maximum argument size. 
Using this script:
```
#!/usr/bin/lua

function write_msg_using_args(msg)
    local fh = io.popen("uclient-fetch -q --post-data='" .. msg .. "' http://127.0.0.1/cgi-bin/test_arg_size -O -")
    fh:read("*a")
    fh:close()
    
    fh = io.open("/tmp/cgi_test")
    local data = fh:read("*a")
    fh:close()
    return data
end

local msg = 'a'
local i = 1
repeat
   assert(msg == write_msg_using_args(msg))
   i = i * 2
   msg = msg .. msg
   print(i)
until( i > 1000000 )

``` 
and this
```
cat /www/cgi-bin/test_arg_size
#!/bin/sh

echo "Content-Type: application/json"
echo

[ "$REQUEST_METHOD" != "POST" ] && exit 22

cat - > /tmp/cgi_test
```

Reply:
> What happens if body is big or if it contains chars like `'` ?
> There is no way to pass it via stdin to `uclient-fetch` if not it would be great to have that feature upstream

To me it looks like it is possible to provide the post data through a pipe (a bit ugly but possible), I included also the utils.shell_quote:

```
local utils = require("lime.utils")
[...]
local cmd = "echo '"..utils.shell_quote(body).."' | uclient-fetch -q -O- --timeout=3 --post-data=- '"..url.."'"
```

Reply:
@ilario How are you testing it? I get the dash `-` written, not the stdin.

Reply:
Yes, I tested on top of OpenWrt 19.07. I just removed a couple of single quotes so that it transfers also data including single quotes.

Reply:
I've tested also in 19.07 and it does not work for me, You can try the script I did (so you don't have to depend on shared-state)

Reply:
also this way you still have the problem of passing the data in the command line, did you ever got the error from shell "too long command"? 

So basically what we need to do here is to write to a temporary file, convince `uclient-fetch` to read post data from there and then delete the file.

Reply:
thanks @spiccinini for the cgi-bin script, I tested using that and I confirm that my solution does not work, sorry for the mistake (and please take #492 seriously)

Reply:
I don't think that `--post-data=-` works, in the code of uclient-fetch it looks like it would simply send `-` as a string rather than reading from stdin.
It would not be hard to add this to uclient-fetch though. Even better would be using libuclient.

Reply:
Yep, I confirmed above that it does not work, I'm going to delete my erroneous comments to reduce confusion.

Reply:
@G10h4ck and me made a patch, you can download it from [patchwork](https://patchwork.ozlabs.org/project/openwrt/patch/20200612142538.GA304638@makrotopia.org/)

```
mkdir package/libs/uclient/patches
curl https://patchwork.ozlabs.org/project/openwrt/patch/20200612142538.GA304638@makrotopia.org/mbox/ >  package/libs/uclient/patches/0001-uclient-fetch-add-option-to-read-POST-data-from-file.patch
```

Reply:
> Yep, I confirmed above that it does not work, I'm going to delete my erroneous comments to reduce confusion.

Please restore that comment :)
It now should work with `uclient-fetch --post-file=/dev/stdin ...` since this is now available in OpenWrt master and openwrt-19.07.